[DROOLS-7372] executable model doesn't resolve bind variable in metho…

tkobayas opened the pull request:
…d call in LHS for property reactivity

- Unit test only

**Ports** 
This is a PR for 7.x
for main ->
for 7.67.x ->
for 7.67.x-blue ->

**JIRA**: 
https://issues.redhat.com/browse/DROOLS-7372


<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] tests</b>

- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

- <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

- <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-main</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-lts</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native</b>

 - for <b>native lts checks</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins run native-lts</b>

- for a <b>specific native lts check</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native-lts</b>
</details>

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

tkobayas commented:
CDB: Not related to this PR
```
[2023-03-24T08:44:48.278Z] [INFO] FAILURE build of project org.kie:business-monitoring-webapp
```

tkobayas commented:
Hi, @pibizza @dupliaka could you review this PR? Thanks!

dupliaka commented on drools-model/drools-model-compiler/src/test/java/org/drools/modelcompiler/PropertyReactivityTest.java:
```diff
@@ -1092,6 +1092,30 @@ public void testExternalFunction2() {
         assertThat(ksession.fireAllRules(3)).isEqualTo(3);
     }
 
+    @Test
```
```suggestion
        @Test
    public void testExternalFunctionWithBindVariableDoNotCauseInfLoop() {
        // DROOLS-7372
        final String str =
                "import " + Fact.class.getCanonicalName() + ";\n" +
                           "import static " + PropertyReactivityTest.class.getCanonicalName() + ".*;\n" +
                           "\n" +
                           "rule R1 when\n" +
                           "    $fact: Fact($id : a == 99999, convertToString($id) == \"BIG\")\n" +
                           "then\n" +
                           "    modify($fact) { setResult(\"OK\") };\n" +
                           "end\n";

        KieSession ksession = getKieSession(str);

        Fact bigString = new Fact();
        bigString.setA(99999);
        bigString.setResult("NG");

        ksession.insert(bigString);
        assertThat(ksession.fireAllRules(3)).isEqualTo(1);
        assertThat(bigString.getResult()).isEqualTo("OK").as("Triggered. A is bound");

        Fact bigStringNE = new Fact();
        bigStringNE.setA(88888);
        bigStringNE.setResult("NG");

        ksession.insert(bigStringNE);
        assertThat(ksession.fireAllRules(3)).isEqualTo(0);
        assertThat(bigStringNE.getResult()).isEqualTo("NG").as("Not triggered. A is bound");

        Fact bigStringNull = new Fact();
        bigStringNull.setResult("NG");

        ksession.insert(bigStringNull);
        assertThat(ksession.fireAllRules(3)).isEqualTo(0);
        assertThat(bigStringNull.getResult()).isEqualTo("NG").as("Not triggered. A is bound. No Exception");
    }
```

dupliaka reviewed (changes requested):
Looks good, I left one suggestion though to expand a bit the coverage of the case. Feel free to comment why it is not needed or please merge the change

tkobayas commented:
Hi @dupliaka ,  Thank you very much for the suggestions and sorry for the late reply.

As you know, I raised a mail thread "Unit test practice", but haven't got much response, so here are my thoughts:

A) Using `as` for `assertThat`

I understand that `as` gives descriptiveness. But I guess we don't want to use `as` for all `assertThat`. So, the criteria to use `as` would likely be "Use `as` when the purpose of the assertion is not clear."

I think `assertThat(ksession.fireAllRules(3)).isEqualTo(1)` is a good candidate to use `as`, so I added its description using `as`, thanks!

B) Expanding test coverage

Suggestion to expand test coverage is always welcome.

However, in this PR, the issue was an infinite loop caused by not resolving a bind variable ('$id'). So the added test cases which don't match the rule are not very relevant to the root problem. In other words, the added test cases would help the coverage of the tested DRL itself, but not the issue which the PR fixed.

Having said that, I have updated the PR. Is it okay for you?

Again, suggestions will be always welcome. Thanks!

tkobayas commented on drools-model/drools-model-compiler/src/test/java/org/drools/modelcompiler/PropertyReactivityTest.java:
```diff
@@ -1092,6 +1092,31 @@ public void testExternalFunction2() {
     }
 
+    @Test
+    public void externalFunctionWithBindVariable_shouldNotCauseInfiniteLoop() {
```
I improved the test method name based on your suggestion, thanks!. Actually some time ago, we had talked about test method name practice.
- No need to start with "test" (because it's obvious with `@Test` annotation)
- `<MethodName>_<ExpectedBehavior>`  (`<MethodName>` could be flexible like `<TestedConditions>`. `_<ExpectedBehavior>` is optional but in this case, it's better to have)

dupliaka replied:
Great! I like it!

tkobayas commented:
Linux - Pull Request: Not related to this PR (Flaky but rare to reproduce)
```
[2023-03-29T06:51:44.190Z] [ERROR] Tests run: 10, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 20.218 s <<< FAILURE! - in org.drools.mvel.integrationtests.KieContainerTest
[2023-03-29T06:51:44.190Z] [ERROR] testIncrementalCompilationSynchronization[KieBase type=CLOUD_IDENTITY](org.drools.mvel.integrationtests.KieContainerTest)  Time elapsed: 20.045 s  <<< ERROR!
[2023-03-29T06:51:44.190Z] org.junit.runners.model.TestTimedOutException: test timed out after 20000 milliseconds
[2023-03-29T06:51:44.190Z]      at java.base@11.0.15/jdk.internal.reflect.Reflection.getCallerClass(Native Method)
[2023-03-29T06:51:44.190Z]      at java.base@11.0.15/java.lang.Class.newInstance(Class.java:576)
[2023-03-29T06:51:44.190Z]      at app//org.drools.reflective.util.ClassUtils.instantiateObject(ClassUtils.java:250)
[2023-03-29T06:51:44.190Z]      at app//org.drools.dynamic.DynamicComponentsSupplier.createConsequenceExceptionHandler(DynamicComponentsSupplier.java:65)
[2023-03-29T06:51:44.190Z]      at app//org.drools.reflective.ComponentsFactory.createConsequenceExceptionHandler(ComponentsFactory.java:45)
```

dupliaka commented:
Do you wait for the approvals from each of the reviewers?

tkobayas commented:
> Do you wait for the approvals from each of the reviewers?

It's sufficient when one of QE reviewers approved. So, @mariofusco , could you merge this PR? Thanks!
